### ISSUE TITLE
dl 

### ISSUE THREAD

**Issue by mqliang** [zhujinghui/Coding-Training#1]
#!/usr/bin/env python
import csv
import os
import tensorflow as tf
import numpy as np
import time
import random
from scipy import stats

LEARNING_RATE = 0.0005  # Learning rate
TRAINING_EPOCHS = 10  # Number of epochs
BATCH_SIZE = 2048 # Batch size
DISPLAY_STEP = 1000  # Display progress each x batches

PADDING_LEN = 14

padding_len = PADDING_LEN
n_input = 40*(2*padding_len+1)
n_hidden_1 = 800  # 1st layer
n_hidden_2 = 640  # 2nd layer
n_hidden_3 = 512  # 3rd layer
n_hidden_4 = 256  # 4th layer
n_hidden_5 = 256  # 5th layer
n_hidden_6 = 160  # 6th layer
n_classes = 138  # output classes

def load_data(path, name):
    return (
        np.load(os.path.join(path, '{}.npy'.format(name)), encoding='bytes'),
        np.load(os.path.join(path, '{}_labels.npy'.format(name)), encoding='bytes')
    )

def expand_Y(Y):
    Y = Y.astype(int)
    YY = np.zeros(shape=(Y.shape[0], n_classes), dtype='uint8')
    for idx, val in enumerate(Y):
        YY[idx][val] = 1
    return YY

def pre_processTrainX(x):
    lookup_table = []
    for i in range(x.shape[0]):
        for j in range(x[i].shape[0]):
            lookup_table.append(0 if len(lookup_table) == 0 else lookup_table[-1]+1)
            if j == x[i].shape[0]-1:
                lookup_table[-1] += padding_len*2
            # if len(lookup_table) == 0:
            #     lookup_table.append(0)
            # if j == x[i].shape[0]-1:
            #     lookup_table.append(lookup_table[-1]+1+padding_len*2)
            # else:
            #    lookup_table.append(lookup_table[-1]+1)

    for i in range(x.shape[0]):
        x[i] = np.pad(x[i], [(padding_len, padding_len), (0, 0)], mode='constant', constant_values=0)
        x[i] = x[i].flatten()

    x = np.concatenate(tuple(x), axis=0)

    return x, lookup_table

class DataBatcher():
    def __init__(self, x, y, lookup_table):
        self.x = x
        self.y = y
        self.lookup_table = lookup_table
        self.batch_size = BATCH_SIZE

        self.xcnt = 0 # item count for x
        self.ycnt = 0 # item count fot y

    def size(self):
        return self.y.shape[0]

    def next_batch(self):
        yy = self.y[self.ycnt*self.batch_size: (self.ycnt+1)*self.batch_size]
        self.ycnt += 1

        xx = np.zeros((self.batch_size, (2*padding_len+1)*40))

        for k in range(self.batch_size):
            idx = self.lookup_table[self.xcnt]
            ctx_data = self.x[idx*40:(idx+2*padding_len+1)*40]
            xx[k] = ctx_data
            self.xcnt += 1

        # using zscore for normalization
        return np.array(stats.zscore(xx, axis=0)), np.array(yy)
        #return xx, np.array(yy)

def pre_processX(x):
    xx = []
    for i in range(x.shape[0]):
        x[i] = np.pad(x[i], [(padding_len, padding_len), (0, 0)], mode='constant', constant_values=0)
    for i in range(x.shape[0]):
        x_flat = x[i].flatten()
        for j in range(0, x[i].shape[0] - 2 * padding_len):
            xx.append(x_flat[j*40:(j+2*padding_len+1)*40])
    # using zscore for normalization
    return stats.zscore(xx, axis=0)
    #return xx

def pre_processY(y):
    yy = np.concatenate(tuple(y), axis=0)
    return expand_Y(yy).astype(float)

def mlp(_X, _weights, _biases):
    layer1 = tf.nn.relu(tf.nn.xw_plus_b(_X, _weights['h1'], _biases['b1']))
    layer2 = tf.nn.relu(tf.nn.xw_plus_b(layer1, _weights['h2'], _biases['b2']))
    layer3 = tf.nn.relu(tf.nn.xw_plus_b(layer2, _weights['h3'], _biases['b3']))
    layer4 = tf.nn.relu(tf.nn.xw_plus_b(layer3, _weights['h4'], _biases['b4']))
    layer5 = tf.nn.relu(tf.nn.xw_plus_b(layer4, _weights['h5'], _biases['b5']))
    layer6 = tf.nn.relu(tf.nn.xw_plus_b(layer5, _weights['h6'], _biases['b6']))
    out = tf.nn.xw_plus_b(layer6, _weights['out'], _biases['out'])
    return out

weights = {
    'h1': tf.get_variable("W1", shape=[n_input, n_hidden_1], initializer=tf.contrib.layers.xavier_initializer()),
    'h2': tf.get_variable("W2", shape=[n_hidden_1, n_hidden_2], initializer=tf.contrib.layers.xavier_initializer()),
    'h3': tf.get_variable("W3", shape=[n_hidden_2, n_hidden_3], initializer=tf.contrib.layers.xavier_initializer()),
    'h4': tf.get_variable("W4", shape=[n_hidden_3, n_hidden_4], initializer=tf.contrib.layers.xavier_initializer()),
    'h5': tf.get_variable("W5", shape=[n_hidden_4, n_hidden_5], initializer=tf.contrib.layers.xavier_initializer()),
    'h6': tf.get_variable("W6", shape=[n_hidden_5, n_hidden_6], initializer=tf.contrib.layers.xavier_initializer()),
    'out': tf.get_variable("out", shape=[n_hidden_6, n_classes], initializer=tf.contrib.layers.xavier_initializer()),
}

biases = {
    'b1': tf.Variable(tf.zeros([n_hidden_1])),
    'b2': tf.Variable(tf.zeros([n_hidden_2])),
    'b3': tf.Variable(tf.zeros([n_hidden_3])),
    'b4': tf.Variable(tf.zeros([n_hidden_4])),
    'b5': tf.Variable(tf.zeros([n_hidden_5])),
    'b6': tf.Variable(tf.zeros([n_hidden_6])),
    'out': tf.Variable(tf.zeros([n_classes]))
}

if __name__ == '__main__':

    # Tf placeholders
    X = tf.placeholder(tf.float32, [None, n_input])
    Y = tf.placeholder(tf.float32, [None, n_classes])

    trainX, trainY = load_data("./data/", "train")
    testX, testY = load_data("./data", "dev")

    trainX, trainX_table = pre_processTrainX(trainX)
    trainY = pre_processY(trainY)

    TRAIN_SIZE = trainY.shape[0]

    testX = pre_processX(testX)
    testY = pre_processY(testY)

    predX = np.load("./data/test.npy", encoding='bytes')
    predX = pre_processX(predX)

    # Build model
    NN = mlp(X, weights, biases)

    # Loss and optimizer
    loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(logits=NN, labels=Y))  # softmax loss
    optimizer = tf.train.AdamOptimizer(learning_rate=LEARNING_RATE).minimize(loss)

    # Accuracy
    correct_prediction = tf.equal(tf.argmax(NN, 1), tf.argmax(Y, 1))
    accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))

    print("Net built successfully...\n")
    print("Starting training...\n")
    # ------------------------------------------------------------------------------
    # Initialize variables
    init_all = tf.global_variables_initializer()

    # Launch session
    config = tf.ConfigProto()
    # config.gpu_options.allow_growth = True

    sess = tf.Session(config=config)
    sess.run(init_all)

    # Training loop
    start_time = time.time()
    # your code
    for epoch in range(TRAINING_EPOCHS):

        random.shuffle(trainX_table) # shuffle

        trainDataBatcher = DataBatcher(x=trainX, y=trainY, lookup_table=trainX_table)

        total_batch = int(TRAIN_SIZE / BATCH_SIZE)
        # Loop over all batches
        for i in range(total_batch):

            batch_xs, batch_ys = trainDataBatcher.next_batch()
            # Fit using batched data
            sess.run(optimizer, feed_dict={X: batch_xs, Y: batch_ys})

            if i % DISPLAY_STEP == 0:
                batch_loss = sess.run(loss, feed_dict={X: batch_xs, Y: batch_ys})
                print ("Epoch: %03d/%03d \t Batch: %03d/%03d loss: %.9f" % (
                epoch, TRAINING_EPOCHS, i, total_batch, batch_loss))
                train_acc = sess.run(accuracy, feed_dict={X: batch_xs, Y: batch_ys})
                print ("Train accuracy: %.3f" % (train_acc))

        # display test accuracy each epoch
        test_acc = sess.run(accuracy, feed_dict={X: testX, Y: testY})
        print ("Test accuracy: %.3f" % (test_acc))

    print ("Training end.\n")

    predY = sess.run(tf.argmax(NN, 1), feed_dict={X: predX})

    csvfile = "./y.csv"
    # Assuming res is a flat list
    with open(csvfile, "w") as output:
        writer = csv.writer(output, lineterminator='\n')
        writer.writerow("id,label\n")
        for idx, val in enumerate(predY):
            writer.writerow([idx, val])

    sess.close()
    print("Session closed!")
